smart-contracts/wasm-transform/src/artifact.rs

```diff
@@ -170,45 +173,57 @@ impl TryFrom<Local> for ArtifactLocal {
 #[derive(Debug, Clone)]
 /// A function which has been processed into a form suitable for execution.
 pub struct CompiledFunction {
-    type_idx:    TypeIndex,
-    return_type: BlockType,
+    type_idx:      TypeIndex,
+    return_type:   BlockType,
     /// Parameters of the function.
-    params:      Vec<ValueType>,
+    params:        Vec<ValueType>,
     /// Number of locals, cached, but should match what is in the
     /// locals vector below.
-    num_locals:  u32,
+    num_locals:    u32,
     /// Vector of types of locals. This __does not__ include function
     /// parameters.
-    locals:      Vec<ArtifactLocal>,
-    code:        Instructions,
+    locals:        Vec<ArtifactLocal>,
+    /// Maximum number of locations needed. This includes parameters,
+    /// locals, and any extra locations needed to preserve values.
+    num_registers: u32,
+    /// The constants in the function.
+    constants:     Vec<i64>,
+    code:          Instructions,
 }
 
 #[derive(Debug)]
 /// A borrowed variant of [CompiledFunction](./struct.CompiledFunction.html)
 /// that does not own the body and locals. This is used to make deserialization
 /// of artifacts cheaper.
 pub struct CompiledFunctionBytes<'a> {
-    pub(crate) type_idx:    TypeIndex,
-    pub(crate) return_type: BlockType,
-    pub(crate) params:      &'a [ValueType],
+    pub(crate) type_idx:      TypeIndex,
+    pub(crate) return_type:   BlockType,
+    pub(crate) params:        &'a [ValueType],
     /// Vector of types of locals. This __does not__ include
     /// parameters.
     /// FIXME: It would be ideal to have this as a zero-copy structure,
     /// but it likely does not matter, and it would be more error-prone.
-    pub(crate) num_locals:  u32,
-    pub(crate) locals:      Vec<ArtifactLocal>,
-    pub(crate) code:        &'a [u8],
+    pub(crate) num_locals:    u32,
+    pub(crate) locals:        Vec<ArtifactLocal>,
+    /// Maximum number of locations needed. This includes parameters,
+    /// locals, and any extra locations needed to preserve values.
+    pub(crate) num_registers: u32,
+    /// The constants in the function.
+    pub(crate) constants:     Vec<i64>, // TODO: Would be better if it was not allocated.
```

Comment:
In reality it does not really matter since in most cases where there are interrupts we allocate things into an OwnedArtifact anyhow.

It would just be more elegant. But I'll remove the TODO.